src/lib/core/CHIPTLVText.cpp

```diff
@@ -0,0 +1,231 @@
+{
+    va_list args;
+
+    va_start(args, aFormat);
+    vprintf(aFormat, args);
+    va_end(args);
+}
+
+CHIP_ERROR Print(TLV::TLVReader &reader, PrinterFunc printerFunc, int tabLevel, bool isFirst)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    if (tabLevel == 0) {
+        err = reader.Next();
+        SuccessOrExit(err);
+    }
+
+    PrintTag(printerFunc, reader.GetTag(), tabLevel, isFirst);
+    
+    switch (reader.GetType()) {
+        case kTLVType_Structure:
+        case kTLVType_Array:
+        case kTLVType_List:
+        {
+            TLVType containerType;
+            const char *openingBrace, *closingBrace;
+
+            if (reader.GetType() == kTLVType_Structure) {
+                openingBrace = "{";
+                closingBrace = "}";
+            }
+            else if (reader.GetType() == kTLVType_Array) {
+                openingBrace = "[";
+                closingBrace = "]";
+            }
+            else {
+                openingBrace = "[[";
+                closingBrace = "]]";
+            }
+
+            if (IsContextTag(reader.GetTag()) || IsProfileTag(reader.GetTag())) {
+                printerFunc("\n");
+                PrintTabs(printerFunc, tabLevel);
+            }
+
+            printerFunc("%s\n", openingBrace);
+
+            err = reader.EnterContainer(containerType);
+            SuccessOrExit(err);
+
+            isFirst = true;
+            while ((err = reader.Next()) == CHIP_NO_ERROR) {
+                err = Print(reader, printerFunc, tabLevel + 1, isFirst);
+                SuccessOrExit(err);
+                isFirst = false;
+            }
+
+            err = reader.ExitContainer(containerType);
+            SuccessOrExit(err);
+
+            printerFunc("\n");
+            PrintTabs(printerFunc, tabLevel);
+            printerFunc("%s", closingBrace);
+            
+            if (tabLevel == 0) {
+                printerFunc("\n");
+            }
+
+            break;
+        }
+
+        case kTLVType_Boolean:
+        {
+            bool v;
+
+            err = reader.Get(v);
+            SuccessOrExit(err);
+
+            printerFunc("%s", (v) ? "true" : "false");
+            break;
+        }
+
+        case kTLVType_SignedInteger:
+        {
+            int64_t v;
+
+            err = reader.Get(v);
+            SuccessOrExit(err);
+
+            printerFunc("%" PRId64 "", v);
+            break;
+        }
+
+        case kTLVType_FloatingPointNumber:
+        {
+            double f;
+
+            err = reader.Get(f);
+            SuccessOrExit(err);
+
+            printerFunc("%.4f", f);
+            break;
+        }
+
+        case kTLVType_UnsignedInteger:
+        {
+            uint64_t v;
+
+            err = reader.Get(v);
+            SuccessOrExit(err);
+
+            printerFunc("%" PRIu64 "", v);
+            break;
+        }
+
+        case kTLVType_ByteString:
+        {
+            uint32_t sz = reader.GetLength();
```

Comment:
This needs some documentation about how we are trusting the length is sane.  And the general setup should be documented to not be OK to run on untrusted TLV that might be malicious.

Reply:
I've switched over to just getting a ByteSpan into the data instead.